runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -83,10 +72,30 @@ protected ILogData removeFromQueue(NavigableSet<Long> queue) {
         if (thisRead != null) {
             // In the case that the immediately requested address (thisRead) is not available,
             // read it along with a batch of subsequent addresses.
-            List<Long> batchRead = getBatch(thisRead, queue);
-            ILogData ld = readRange(thisRead, batchRead);
 
-            if (queue == getCurrentContext().readQueue) {
+            ILogData ld;
+
+            try {
+                // The read range has a special behaviour, it attempts to read the address of interest,
+                // and in the case it is not in the cache, it reads a batch of addresses
+                // that will be further required, returning only the data for the address of interest.
+
+                // Because addresses are requested in increasing order and system relies on prefix trim,
+                // we force trimmedExceptions to be thrown by lower layers, and handle ignoreTrimmed at this layer.
+                // Note that lower layers will cache the valid entries, to optimize read performance.
+                ld = readRange(thisRead, queue, false);
+            } catch (TrimmedException te) {
+                if (!options.ignoreTrimmed) {
+                    throw te;
+                }
+
+                log.warn("removeFromQueue[{}]: ignoring trimmed addresses {}", this, te.getTrimmedAddresses());
+                // Ignore trimmed address, remove trimmed addresses and get next from queue
+                te.getTrimmedAddresses().forEach(address -> queue.remove(address));
+                return removeFromQueue(queue);
+            }
+
+            if (queue == getCurrentContext().readQueue && ld != null) {
```

Comment:
readRange return value is marked @Nonnull, `ld != null` is a redundant check.

Reply:
actually this annotation is only informational, it's the caller's responsibility to check it is not null.